### ISSUE TITLE
RS232 control of digital scale

### ISSUE THREAD

**Issue by sgbaird** [AccelerationConsortium/ac-training-lab#20]
Using https://www.waveshare.com/wiki/Pico-2CH-RS232 and [US Solid 3kg scale](https://www.amazon.ca/U-S-Solid-Electronic-AC100V-240V-Measurement/dp/B09X12DCNG/)

EDIT: one Pico-2CH-RS232 is available (in 3D printer room). Both the scale and the adapter that came with the RS232 module are male, so I also ordered  [F-F adapters](https://www.amazon.ca/gp/product/B0CXLZ9F4M/ref=ppx_yo_dt_b_asin_title_o00_s00?ie=UTF8&th=1). The same applies for the A&D scale.

I had trouble finding the instructions/manual for the US Solid scale online. Here is a scan of a paper copy:

[uss-dbs-series-operation-manual.pdf](https://github.com/user-attachments/files/16936627/uss-dbs-series-operation-manual.pdf)

The directions are somewhat sparse, perhaps because continuous output is essentially the only option (compared with A&D which has a lot more functionality exposed).

Related microcourse module: https://ac-microcourses.readthedocs.io/en/latest/courses/robotics/3.2-serial-communication.html

**Comment by evelynnesher** [comment]
![fullshot1](https://github.com/user-attachments/assets/01974060-14da-492d-8c5c-635a950f6025)
![fullshot2](https://github.com/user-attachments/assets/dca12ba6-b371-468d-a080-457daacda49b)
![genderch](https://github.com/user-attachments/assets/1afd09a2-08fe-4b27-a88f-d6840cb6af2a)
![rasp1](https://github.com/user-attachments/assets/84513bb3-69f0-48cc-ac6b-e5f6af68baf1)
Trying to connect the scale to the raspberry pi + base but can't seem to get it to connect.

**Comment by sgbaird** [comment]
Hi @evelynnesher, here is the [documentation for the PICO-2CH-RS232 module](https://www.waveshare.com/wiki/Pico-2CH-RS232). Connecting the scale directly to the Pico W won't work. You also won't need to user the Maker Pi Pico base. Just the Pico W and the RS232 module (I think it's already plugged in).

The resources from the first post should also help: https://github.com/AccelerationConsortium/ac-training-lab/issues/20#issue-2503809952

**Comment by evelynnesher** [comment]
@sgbaird I think I don't have a Industrial RS232 to 3.3V TTL Converter.

**Comment by sgbaird** [comment]
> @sgbaird I think I don't have a Industrial RS232 to 3.3V TTL Converter.

Correct, I'm mostly referring to the code snippet. You have this one https://www.waveshare.com/pico-2ch-rs232.htm (an alternative to the one you linked).

**Comment by evelynnesher** [comment]
![setup2](https://github.com/user-attachments/assets/40778169-3d60-4fe6-aca2-0f34c8828e0a)
![wire setup](https://github.com/user-attachments/assets/06600a66-023b-4492-a639-c5ef2df0a78e)
Did I set this up correctly? Still having connection issues

**Comment by sgbaird** [comment]
I think that's correct. Could you describe the connection issue in more detail?

**Comment by evelynnesher** [comment]
When all the parts connect, it is not reading any data.
![image](https://github.com/user-attachments/assets/397ee581-5462-4513-84a8-5ef55d6ea239)
Specifically, I make sure to check the scale is connected but uart.any() is always empty.
I also make sure to follow the instructions on page 7 of the scale operations manual, data output settings.

**Comment by sgbaird** [comment]
Can you share the full code snippet that you're running? Use three backticks to apply code formatting

**Comment by evelynnesher** [comment]
The following code from module 3.2 of the microcourse:
```
import machine
import utime

# Initialize UART
uart = machine.UART(1, baudrate=9600, tx=machine.Pin(4), rx=machine.Pin(5))

# Send the "Q" command and read response
uart.write(b"Q\r\n")
utime.sleep(0.1)
if uart.any():
    print(uart.read().decode("utf-8"))

# Continuous reading loop
while True:
    if uart.any():
        print(uart.read().decode("utf-8"))
    utime.sleep(1)

**Comment by sgbaird** [comment]
You can remove the "Q" command. That doesn't apply to the US solid scale. It may not affect anything, but will make debugging easier. 

Assuming you've set the scale up in continuous output mode per the instruction manual, it may be worth using the uartassist software that was mentioned and connect directly to your computer, however you'll need a different type of cable. I think I have the required connectors in the office. I'll set those aside for you to pick up (likely tomorrow).

**Comment by sgbaird** [comment]
Actually, it looks like you have it connected into channel 0, however you're specifying the pins for channel 1 I think. Check the pin legend on the Pico 2ch rs232 wiki.

**Comment by evelynnesher** [comment]
@sgbaird 
I am trying to correctly set up the scale to send continuous output. I follow the continuous output instructions (pasted below/ found on page 7 of the instructions manual), once I get to the last step and I click TARE, it sends the display to C06-0. I provided the reference for C06 displays.

![image](https://github.com/user-attachments/assets/c4a208e0-b5cd-48c5-bf84-c136d4003532)

![image](https://github.com/user-attachments/assets/855358d1-0165-41d8-abf5-02ba6a637715)

I had ChatGPT write Thonny code that checks if the scale is connected. I believe my issue either lies in the continuous output or the physical setup. 
```import machine
import utime

# Initialize UART1 on pins GP4 (TX) and GP5 (RX)
uart = machine.UART(1, baudrate=9600, tx=machine.Pin(4), rx=machine.Pin(5))

def check_scale_connection():
    print("Checking if the scale is connected...")

    # Try reading data from the UART
    if uart.any():  # If there's data available on the UART
        response = uart.read()  # Read the data
        if response:
            try:
                # Try to decode the response as UTF-8
                decoded_data = response.decode("utf-8").strip()
                print("Scale connected. Data received:", decoded_data)
            except UnicodeDecodeError:
                # If the data cannot be decoded, show it in raw form
                print("Scale connected. Non-text data received:", response)
        else:
            print("Scale connected, but no data received.")
    else:
        print("No data received. The scale may not be connected or may not be sending data.")

# Run the check in a loop to continuously check if the scale is connected
while True:
    check_scale_connection()
    utime.sleep(2)  # Check every 2 seconds

**Comment by sgbaird** [comment]
Thanks for the info! I'm not sure what the issue is. Happy to take a look when you're nearby. Otherwise, you could take a video of the full process and I can look over it.

**Comment by evelynnesher** [comment]
https://github.com/user-attachments/assets/60bff6e9-e8eb-4627-bbc4-e17e1bfe4ec2

I added a video of me following the manual instructions. As for the physical connection, it is the same as the picture I added yesterday except it is in channel 1 instead of 0 to correspond with the code. I have two adaptations of the F-F adapter. When I take the screws out it falls apart/ doesn’t stay inside the port in the scale so I just play around with it until it stays.

![IMG_2139](https://github.com/user-attachments/assets/8b21fc85-8e33-4c3a-81e8-c05cfe4307a7)

https://github.com/user-attachments/assets/ab9fdf10-f28d-4be1-8f42-a9c7c035b08f

**Comment by sgbaird** [comment]
It's different from what the instructions say, but maybe worth trying a long press of the tare at the end, or ignoring the last step of pressing tare. Directly connecting to a computer with a DB9 to USB-A cable (in one of the boxes in the room with the 3D printer), and using the uartassist software will likely help with the troubleshooting. We also have a second type of scale that probably has better documentation.

**Comment by sgbaird** [comment]
Aside: Some additional Pico 2-channel RS232 modules came in 

![PXL_20241001_154949742.jpg](https://github.com/user-attachments/assets/c837bb15-bff1-443c-993d-9ab718d32d26)

**Comment by evelynnesher** [comment]
I’ve gone through every combination of buttons, so the long TARE at the end isn’t the solution. I’ll stop by around 3 tomorrow to pick up the new materials. Will you be available then? If so, we can take a look together; if not, I’ll keep trying to sort it out on my own.

**Comment by sgbaird** [comment]
Here's the cable that you should be able to use to connect directly to your computer for troubleshooting. It's in the room with a 3D printer

![PXL_20241001_201422446.jpg](https://github.com/user-attachments/assets/5dc182ca-18ce-4e3d-95ce-86e308e1c24c)

There is a second scale that may be worth troubleshooting with. This one is more sensitive and needs to stay in the office. Instructions are also different, but I think there are links on the micro course website. This is one that we know can do serial communication.

![PXL_20241001_204033333.jpg](https://github.com/user-attachments/assets/d9734b59-4405-4431-974d-3b3768077eb3)

**Comment by evelynnesher** [comment]
Thank you! I’d like to start with the first one you provided. If that doesn’t work, I’ll come back and try the second scale.

**Comment by evelynnesher** [comment]
![image](https://github.com/user-attachments/assets/8e0614ef-e76d-4066-8402-5a80a884a26c)
UartAssist Software worked. Worked with Channel COM9 with the baud rate, parity, data bits, stop bits, and flow control set to the values specified by the scale manual.

**Comment by sgbaird** [comment]
Amazing! Narrows things down for sure.

**Comment by sgbaird** [comment]
And these were the correct instructions?

https://github.com/AccelerationConsortium/ac-training-lab/issues/20#issuecomment-2375510617

**Comment by evelynnesher** [comment]
These were the instructions available for using the software
![image](https://github.com/user-attachments/assets/1f54e972-3619-4799-bbda-cb1233c6a579)
To use the UartAssist Software, you configure the serial options (top left) then click open and it gives you continuous input.

**Comment by evelynnesher** [comment]
I'm trying everything to debug the issue. Could the problem be related to the fact that the scale's output is 110V/220V while the Pico operates at 3.3V~5V? Could this voltage difference be causing the communication issue?
Scale technical parameters:
![image](https://github.com/user-attachments/assets/e7eb360f-963c-486a-b5d9-cbc0e2eff96e)
Pico-2CH-RS232 Specifications:
![image](https://github.com/user-attachments/assets/97d17193-5f6b-43ae-a48c-9e674ac03960)

**Comment by evelynnesher** [comment]
As part of my debugging efforts, I wrote some code to check the connection with the scale. The uart.any() method checks for available data in the UART and returns the number of bytes present in the receive buffer. However, it hasn’t provided any clarity, as it consistently returns False/zero. This means that there is no data available (I’m not sure if you already know this, but I wanted to clarify since you mentioned you're not familiar with the RS232 module). Any suggestions/next steps?
![image](https://github.com/user-attachments/assets/0c7e0706-b8c0-4627-964b-7bb2608a40ad)

```
import machine
import utime

# Initialize UART
#uart = machine.UART(1, baudrate=9600, tx=machine.Pin(4), rx=machine.Pin(5))
uart = machine.UART(1, baudrate=9600, tx=machine.Pin(4), rx=machine.Pin(5), bits=8, parity=None, stop=1)

#pins 4 and 5, 8 and 9 work
# Function to check if the scale is connected
def check_scale_connection():

    utime.sleep(0.1)  
    
    if uart.any():  # Check if there's any response
        response = uart.read()  # Read the response
        if response:
            print("Scale connected. Response:", response.decode("utf-8").strip())
        else:
            print("Scale is not responding.")
    else:
        print("No response from the scale. Check connections.")

# Check if the scale is connected
check_scale_connection()

**Comment by sgbaird** [comment]
Various resources:


Demo code from the "Download Example" section from [the wiki](https://www.waveshare.com/wiki/Pico-2CH-RS232), using https://files.waveshare.com/upload/5/5a/Pico_2CH_RS232_Code.7z, which can be downloaded manually and viewed (Windows seemed to do fine opening it up even in the archived state, not sure if that's because I did something special a while back to get it to work with 7z, or if it's a default for Windows now). The following Python example is given:

```python
from machine import UART, Pin
import time

uart1 = UART(1, baudrate=9600, tx=Pin(4), rx=Pin(5))

uart0 = UART(0, baudrate=9600, tx=Pin(0), rx=Pin(1))

txData = b'RS232 receive test...\r\n'
uart0.write(txData)
time.sleep(0.1)
while True:
        #rxData = bytes()
    while uart0.any() > 0:    #Channel 0 is spontaneous and self-collecting
        rxData0 = uart0.read()
        uart0.write("{}".format(rxData0.decode('utf-8')))
        print(rxData0)
        if(uart0.any()==0):
            uart0.write("\r\n")
    while uart1.any() > 0:   #Channel 1 is spontaneous and self-collecting
        rxData1 = uart1.read()
        uart1.write("{}".format(rxData1.decode('utf-8')))
        print(rxData1)
        if(uart1.any()==0):
            uart1.write("\r\n")
```

Some other code examples:
- [General-purpose UART example code by Waveshare](https://github.com/waveshareteam/Pico_MircoPython_Examples/blob/main/05_UART/UART.py)
- [CircuitPython example of using the Waveshare Pico-2CH-RS232 module (user project)](https://github.com/chabala/pi-pico-serial-web-api/blob/e3c89d44ea4975709d2528777ea8ce86ac57d038/CIRCUITPY/code.py#L37)

**Comment by sgbaird** [comment]
I think the issue may actually be the use of the female-female adapter, in that it flips the connections from what is expected.. this was my bad. I'm planning to get something like the following so we can have more control:

Futheda 5PCS D-SUB DB9 RS232 Interface Breakout Board Connector 9-pin 2-Row Female RS-232 Serial Port Solderless Terminal Block Adapter with Positioning Nuts https://a.co/d/6LEEta6

Likewise, the existing connector could be modified, but maybe better to just wait and go with the one above.

**Comment by evelynnesher** [comment]
### Submitted the following ticket to Pico-2CH-RS232 Wiki page on October 11, 2024.

Hi,
I hope you're doing well.

I'm working on connecting a Pico-2CH-RS232 module to a U.S. Solid Digital Scale. I'm following the RS232 module instructions for continuous output as outlined in the scale's manual, aiming to have the Pico collect and manage the data. The digital scale is connected to Channel 1 on the Pico Base using a Dupont 3-pin to DB-9 wire cable. We've confirmed that the scale works by testing the continuous output with UartAssist software.

However, the main issue we're facing is that, when using MicroPython on Thonny, we cannot read any connection from the scale. After following the continuous output instructions from the scale manual, nothing happens. We have also tried the basic example demo code from the "Download Example" section from the wiki, using https://files.waveshare.com/upload/5/5a/Pico_2CH_RS232_Code.7z

Here’s the code I’m currently using, which continuously prints "No response from the scale. Check connections":

```python
import machine
import utime

# Initialize UART
# uart = machine.UART(1, baudrate=9600, tx=machine.Pin(4), rx=machine.Pin(5))
uart = machine.UART(1, baudrate=9600, tx=machine.Pin(4), rx=machine.Pin(5), bits=8, parity=None, stop=1)

# Function to check if the scale is connected
def check_scale_connection():
while True:
if uart.any(): # Check if there's any response
response = uart.read() # Read the response
if response:
print("Scale connected. Response:", response.decode("utf-8").strip())
else:
print("Scale is not responding.")
else:
print("No response from the scale. Check connections.")

# Check if the scale is connected
check_scale_connection()
```

Here’s the GitHub thread where we've outlined all the troubleshooting steps we've taken so far:
https://github.com/AccelerationConsortium/ac-training-lab/issues/20

Please let me know if you need further information or have any suggestions on how I can proceed.

Best regards,
Evelyn

**Comment by evelynnesher** [comment]
https://github.com/user-attachments/assets/9da69e02-e084-4394-aefa-b2872de8c110

![IMG_2538](https://github.com/user-attachments/assets/5a7f6627-7e1d-4f9e-a25b-22087e5c47d8)
![IMG_2537](https://github.com/user-attachments/assets/0b4fd202-910c-45e9-adf1-f356bda25993)

I got it working! I used the Female DB9 breakout board adapter with three male-to-male jumper wires. I connected GND to GND, and TX on the breakout board to RX on the Pico. I didn’t need to use the DB9 to multicore cable. I also sent a video showing the continuous output. I’ll attach the code I used as well.

**Comment by evelynnesher** [comment]
![IMG_2546](https://github.com/user-attachments/assets/89306266-2d32-492c-a05e-853efc656769)

By using the new adapter, I realized the connection issue wasn’t with the Female-Female adapter but with the TX-RX connection. I used the male-male jumper cables with our original setup instead of directly plugging in the DB9 to TX1, RX1 and GND of the microcontroller. I connected the jumper wires as shown in the picture: black to black and red to yellow. 
Now, we have two working setups, both very easy to assemble. The only issue with this one is that the Female-Female adapter tends to fall apart because we had to remove the nuts that hold it together to fit it to the scale. No screwdriver was needed for this setup, while one was required for the one mentioned above. The connection is stable in both setups and the speed at which the scale sends data seems to be the same with both setups, but I’ll test it to confirm. When securing the jumper wires into the terminal blocks (screw terminal connectors), the wires may bend slightly. This happens in both setups, but in this one, it’s only on one side, while in the previous setup, it happens on both. However, it doesn’t affect functionality.

**Comment by sgbaird** [comment]
Way to go! That's great you figured it out on both setups. Great progress.

**Comment by sgbaird** [comment]
Let's ditch the F-F adapter setup. I think sticking with the terminal breakout is fine.

I wouldn't worry much about the data transfer speed. Instead, the next priority will be setting up MQTT and a hugging face spaces web app for it.

This will be very similar to the fan control demo, except without the "send command" aspect. Will share some links.

**Comment by sgbaird** [comment]
Now that you have the scale working, can you start working on an MQTT + gradio web app implementation for one way streaming of the scale data?

**Comment by evelynnesher** [comment]
Definitely, I’ve started already. You want me to use Gradio instead of Streamlit?

**Comment by sgbaird** [comment]
Great, and yes - it would be best to switch over (see https://github.com/AccelerationConsortium/ac-training-lab/issues/77#issuecomment-2429034334).

**Comment by evelynnesher** [comment]
I successfully established the MQTT connection and will now start implementing Gradio. It's currently set up to send the following data:

```json
{
    "weight": "15.5g",
    "time": "2024-10-30 12:34:56",
    "device_id": "ambiguous-goose"
}
```

On HiveMQ, I can see the following information organized under the columns **Message**, **Topic**, and **QoS**:
![image](https://github.com/user-attachments/assets/4c2f10cd-ef80-4828-bc9b-c7ee61ea6d4b)

For Hugging Face Spaces, would you like a button to turn continuous data on and off, and should the scale send just the weight or include the timestamp and other information? Should the output appear below the button, or would you prefer it to be displayed elsewhere?

**Comment by sgbaird** [comment]
Nice! That's great. I think weight and timestamp works, and you can use a unique ID from the Pico W in the topic structure and leave off the device ID. Something like `<US Solid product ID>/<Pico ID>`.

For id, see https://github.com/sparks-baird/self-driving-lab-demo/blob/3d45417f1832679f84044f52e3c64daeb5437ca3/src/public_mqtt_sdl_demo/main.py#L63. For UTC timestamp, see (https://github.com/sparks-baird/self-driving-lab-demo/blob/3d45417f1832679f84044f52e3c64daeb5437ca3/src/public_mqtt_sdl_demo/lib/sdl_demo_utils.py#L188 and https://github.com/sparks-baird/self-driving-lab-demo/blob/3d45417f1832679f84044f52e3c64daeb5437ca3/src/public_mqtt_sdl_demo/main.py#L106) (i.e., you should synchronize the time via ntptime before calling).

For Hugging Face, I think as soon as it starts up, then start streaming the weigh data. You will want to try to avoid re-instantiating the MQTT client within a session.

Also, I recommend using some variation of the plotting code from https://huggingface.co/spaces/AccelerationConsortium/fan-control

**Comment by evelynnesher** [comment]
![image](https://github.com/user-attachments/assets/1a39e31e-a55e-475f-99cd-3b3b3df47b32)
![image](https://github.com/user-attachments/assets/8a142033-dcea-407d-a623-419a846e4c84)

I’ve completed the Gradio app in a personal space on my account. I made two versions: one displays data when you click the refresh button, and the other updates continuously. Let me know what changes to make. I’ll also need access to the Acceleration Consortium space so I can implement it there.

Currently, I run main.py on Thonny to collect the data and send it to HiveMQ, while I run app.py on Hugging Face Spaces to create the graph that retrieves data from HiveMQ.

**Comment by sgbaird** [comment]
> Needs Refresh: ![image](https://private-user-images.githubusercontent.com/159678938/382381520-18dd4f68-38af-40e3-9817-83d622bbeb80.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA0OTEwNzcsIm5iZiI6MTczMDQ5MDc3NywicGF0aCI6Ii8xNTk2Nzg5MzgvMzgyMzgxNTIwLTE4ZGQ0ZjY4LTM4YWYtNDBlMy05ODE3LTgzZDYyMmJiZWI4MC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEwMVQxOTUyNTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mNGZjNWFjYWMzMDYwYmQ2NGY1OWE0YTNiMGMzZjFmZGFkN2NlNWRiMTY1OTNmMDJlNzc5OWE2ZTI0NjJhZTM1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.GksqPF2mu8UFDVGDLwwa3dpVpuM7EH0XWeoEL_SPiaI)![image](https://private-user-images.githubusercontent.com/159678938/382372765-1a39e31e-a55e-475f-99cd-3b3b3df47b32.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA0OTEwNzcsIm5iZiI6MTczMDQ5MDc3NywicGF0aCI6Ii8xNTk2Nzg5MzgvMzgyMzcyNzY1LTFhMzllMzFlLWE1NWUtNDc1Zi05OWNkLTNiM2IzZGY0N2IzMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEwMVQxOTUyNTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04MjNlNjQ2NmE3MTc0MTNmNGZjMTFkMTlkN2JiZjM3MWM4MzE3NjJmZjYxMDcxMWIzMjE4YzJiYTg1YmRiNjIwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Sy2O-Ez2-CkvExr9Qyu4sUENHZnYIuC4ahtTsvt5t2E) Continuous: ![image](https://private-user-images.githubusercontent.com/159678938/382381520-18dd4f68-38af-40e3-9817-83d622bbeb80.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA0OTEwNzcsIm5iZiI6MTczMDQ5MDc3NywicGF0aCI6Ii8xNTk2Nzg5MzgvMzgyMzgxNTIwLTE4ZGQ0ZjY4LTM4YWYtNDBlMy05ODE3LTgzZDYyMmJiZWI4MC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEwMVQxOTUyNTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mNGZjNWFjYWMzMDYwYmQ2NGY1OWE0YTNiMGMzZjFmZGFkN2NlNWRiMTY1OTNmMDJlNzc5OWE2ZTI0NjJhZTM1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.GksqPF2mu8UFDVGDLwwa3dpVpuM7EH0XWeoEL_SPiaI) ![image](https://private-user-images.githubusercontent.com/159678938/382375056-8a142033-dcea-407d-a623-419a846e4c84.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA0OTEwNzcsIm5iZiI6MTczMDQ5MDc3NywicGF0aCI6Ii8xNTk2Nzg5MzgvMzgyMzc1MDU2LThhMTQyMDMzLWRjZWEtNDA3ZC1hNjIzLTQxOWE4NDZlNGM4NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEwMVQxOTUyNTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00NWZmMDA0ZTk5OGU4ZDRmZmJlYmNiZTdiM2I3MTVmNDdlYzU3YzU1NGRlNGNiYmI4ZjllYTYyNTMyZDRiMjlmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.mRZADN19-x8mvECb_n_JcvAUiFa_H3tsNpfjQvQgKOE)
> 
> I’ve completed the Gradio app in a personal space on my account. I made two versions: one displays data when you click the refresh button, and the other updates continuously. The continuous one currently updates slowly, but I'm working on sending real-time updates.
> 
> Let me know what changes to make. I’ll also need access to the Acceleration Consortium space so I can implement it there.
> 
> Currently, I run main.py on Thonny to collect the data and send it to HiveMQ, while I run app.py on Hugging Face Spaces to create the graph that retrieves data from HiveMQ.

Nice! Can you provide links to these two spaces?

**Comment by evelynnesher** [comment]
https://huggingface.co/spaces/evelyn-nesher19/scale-test-continous
https://huggingface.co/spaces/evelyn-nesher19/scale-test-refresh
I made distinct visual differences between them so we can choose which best suits our implementation. For the continuous version, we may want to add a screen that shows the exact weight, as it’s difficult to read precise values when the weight fluctuates significantly. 
If the graph isn’t connecting, it’s likely because the Raspberry Pi stopped sending information.

**Comment by evelynnesher** [comment]
Continuously updating graph : https://huggingface.co/spaces/evelyn-nesher19/scale-test-continous
This graph updates smoothly in real-time and includes a refresh button for manual updates. 

Manual Refresh Needed: https://huggingface.co/spaces/evelyn-nesher19/scale-test-refresh
This version has a "connection status" button and requires manual refreshing to display the latest data.

**Comment by sgbaird** [comment]
Maybe use a sliding window for the data, with a history of ~30 seconds? I don't think a manual refresh is needed in this case. Maybe have a look at the main.py for the fan control demo in the AC training lab repo (nested under `src`). You can reduce the time sampling so that data points get sent every 2-3 seconds for example.

Overall looks good though!

**Comment by evelynnesher** [comment]
I just wanted to clarify the suggestion about using a sliding window for the data. Does this mean we'll only display the data using the sliding window, or should I keep the graph showing a few seconds of data and have the sliding window below it to show the history for the past 30 seconds?

**Comment by evelynnesher** [comment]
![image](https://github.com/user-attachments/assets/294e1fc8-9a57-49c9-9684-e875a92f2e39)
I've embedded the space into Gather Town!

**Comment by sgbaird** [comment]
Nice! Looks great 😊 

I noticed something when trying to click connect:

```python
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/gradio/queueing.py", line 624, in process_events
    response = await route_utils.call_process_api(
  File "/usr/local/lib/python3.10/site-packages/gradio/route_utils.py", line 323, in call_process_api
    output = await app.get_blocks().process_api(
  File "/usr/local/lib/python3.10/site-packages/gradio/blocks.py", line 2015, in process_api
    result = await self.call_function(
  File "/usr/local/lib/python3.10/site-packages/gradio/blocks.py", line 1562, in call_function
    prediction = await anyio.to_thread.run_sync(  # type: ignore
  File "/usr/local/lib/python3.10/site-packages/anyio/to_thread.py", line 56, in run_sync
    return await get_async_backend().run_sync_in_worker_thread(
  File "/usr/local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 2441, in run_sync_in_worker_thread
    return await future
  File "/usr/local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 943, in run
    result = context.run(func, *args)
  File "/usr/local/lib/python3.10/site-packages/gradio/utils.py", line 865, in wrapper
    response = f(*args, **kwargs)
  File "/home/user/app/app.py", line 46, in connect_mqtt
    client.tls_set()
  File "/usr/local/lib/python3.10/site-packages/paho/mqtt/client.py", line 1302, in tls_set
    self.tls_set_context(context)
  File "/usr/local/lib/python3.10/site-packages/paho/mqtt/client.py", line 1193, in tls_set_context
    raise ValueError('SSL/TLS has already been configured.')
ValueError: SSL/TLS has already been configured.
```

**Comment by evelynnesher** [comment]
Fixed! I don't have access to edit AC Hugging Face Spaces Collections, so can't place under "Hardware Control" Collection. Ready to set it up at the office and start my next task.

**Comment by sgbaird** [comment]
Nice! If there are any updates to `main.py` on the microcontroller, do you mind making a PR? If not, you can ignore. The Pico W will need the Nokia credentials, and you can use the 248 broker and username/password (the public test credentials) for now. We may switch to a different one a bit later.

**Comment by sgbaird** [comment]
Perhaps the next task will be the autotrickler, since that carries over the RS232 experience and is a natural extension.

**Comment by sgbaird** [comment]
Great! Could you comment briefly on how you implemented this? This relates to https://github.com/AccelerationConsortium/ac-training-lab/issues/35, which we're still not decided on in terms of a convention we want to follow.

**Comment by evelynnesher** [comment]
There are two global variables: connection_status and data_status which store the current MQTT connection status and the state of data reception, respectively. If you need to track other statuses, like a connection for another component or a different piece of data, you can easily add more variables and follow the same approach.

A mutex (status_lock) is used to make sure that only one part of the program can update or read the status at a time. This is important because the program has multiple threads running at once (e.g., the MQTT client handling incoming messages and the Gradio app updating the user interface). Without the mutex, both threads might try to change or read the status at the same time, which could cause errors or confusing results. The mutex ensures that these updates happen one at a time, keeping everything in sync.

The get_status function formats and returns the current status as a string while ensuring thread safety.

In the Gradio app, the status is displayed using a gr.Markdown component (status_md). This component is refreshed every second (every=1), ensuring that the status is kept up-to-date for the user without manual refreshes.

Long story short: The get_status function is called every second to retrieve the latest values for connection_status and data_status, and the Gradio UI is refreshed accordingly to show the updated status. 

The current setup causes the UI to "flash" with each refresh. To fix this, I plan to adjust the implementation so that the UI only refreshes when the connection_status changes, instead of refreshing every second.

**Comment by sgbaird** [comment]
Nice! A simple workaround for now may be to just increase to 30 s. When a user first goes to the page is the most important time to see the status in my mind (especially if it's not connected or receiving data).

**Comment by sgbaird** [comment]
Nice! You can use one of the 5V power adapters any plug it in normally instead of the built in USB spot. Thanks!

**Comment by sgbaird** [comment]
Anything left for this one?

**Comment by evelynnesher** [comment]
Good to close!
